File: test_extension.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
CTF: Discord Token Extractor Extension Validation
Automated tests for token extraction methods
"""

import json
import os
import re
import sys
from pathlib import Path

# Test Results
TEST_RESULTS = {
    'passed': [],
    'failed': [],
    'warnings': [],
}

def log_success(msg):
    print("[OK] " + msg)

def log_error(msg):
    print("[NG] " + msg)

def log_warn(msg):
    print("[WARN] " + msg)

def log_info(msg):
    print("[INFO] " + msg)

# ============================================
# Test 1: manifest.json validation
# ============================================
def test_manifest():
    log_info('Test 1: manifest.json structure validation')
    try:
        manifest_path = Path(__file__).parent / 'manifest.json'
        with open(manifest_path, 'r', encoding='utf-8') as f:
            manifest = json.load(f)

        required = ['manifest_version', 'name', 'version', 'action', 'permissions']
        for field in required:
            if field not in manifest:
                raise ValueError("Missing field: " + field)

        if manifest['manifest_version'] != 3:
            raise ValueError("manifest_version must be 3")

        if 'scripting' not in manifest['permissions']:
            raise ValueError('Missing "scripting" permission')

        if 'clipboardWrite' not in manifest['permissions']:
            raise ValueError('Missing "clipboardWrite" permission')

        if 'host_permissions' not in manifest or \
           not any('discord.com' in p for p in manifest.get('host_permissions', [])):
            raise ValueError('Missing Discord host permission')

        TEST_RESULTS['passed'].append('manifest.json structure valid')
        log_success('manifest.json valid')
    except Exception as e:
        TEST_RESULTS['failed'].append("Manifest validation: " + str(e))
        log_error("manifest.json validation failed: " + str(e))

# ============================================
# Test 2: popup.html validation
# ============================================
def test_popup_html():
    log_info('Test 2: popup.html structure validation')
    try:
        html_path = Path(__file__).parent / 'popup.html'
        with open(html_path, 'r', encoding='utf-8') as f:
            html = f.read()

        if 'extractButton' not in html:
            raise ValueError('Missing extractButton element')
        if 'popup.js' not in html:
            raise ValueError('Missing popup.js script')
        if 'Discord' not in html and 'discord' not in html:
            raise ValueError('Missing Discord label text')

        TEST_RESULTS['passed'].append('popup.html structure valid')
        log_success('popup.html valid')
    except Exception as e:
        TEST_RESULTS['failed'].append("HTML validation: " + str(e))
        log_error("popup.html validation failed: " + str(e))

# ============================================
# Test 3: popup.js function validation
# ============================================
def test_popup_js():
    log_info('Test 3: popup.js function validation')
    try:
        js_path = Path(__file__).parent / 'popup.js'
        with open(js_path, 'r', encoding='utf-8') as f:
            js = f.read()

        required = ['extractTokenFromTab', 'extractToken']
        for fn in required:
            if ('function ' + fn) not in js:
                raise ValueError("Missing function: " + fn)

        methods = [
            ('webpackChunk', r'window\.webpackChunkdiscord_app'),
            ('localStorage', r'localStorage\.getItem|localStorage\.token'),
            ('sessionStorage', r'sessionStorage'),
            ('ReactFiber', r'__react'),
        ]

        implemented = [name for name, pattern in methods if re.search(pattern, js)]
        log_info("Extraction methods: " + ', '.join(implemented))

        if len(implemented) < 2:
            TEST_RESULTS['warnings'].append("Only " + str(len(implemented)) + " extraction methods found")
            log_warn("Only " + str(len(implemented)) + " extraction methods found")

        TEST_RESULTS['passed'].append('popup.js functions present and valid')
        log_success('popup.js valid')
    except Exception as e:
        TEST_RESULTS['failed'].append("JavaScript validation: " + str(e))
        log_error("popup.js validation failed: " + str(e))

# ============================================
# Test 4: Token extraction logic test
# ============================================
def test_token_extraction():
    log_info('Test 4: Token extraction logic validation')
    try:
        js_path = Path(__file__).parent / 'popup.js'
        with open(js_path, 'r', encoding='utf-8') as f:
            js = f.read()

        if 'function extractToken()' not in js:
            raise ValueError("extractToken function not found")

        if 'try' not in js or 'catch' not in js:
            raise ValueError("Error handling not implemented")

        patterns = [
            (r'webpackChunkdiscord_app.*?push', 'webpack chunk method'),
            (r'localStorage\.getItem', 'localStorage method'),
            (r'sessionStorage', 'sessionStorage method'),
        ]

        found = [desc for pattern, desc in patterns if re.search(pattern, js)]
        if len(found) < 2:
            raise ValueError("Too few extraction methods: " + str(found))

        TEST_RESULTS['passed'].append('Token extraction logic valid')
        log_success('Token extraction logic validated')
    except Exception as e:
        TEST_RESULTS['failed'].append("Token extraction test: " + str(e))
        log_error("Token extraction test failed: " + str(e))

# ============================================
# Test 5: Permission validation
# ============================================
def test_permissions():
    log_info('Test 5: Chrome permission validation')
    try:
        manifest_path = Path(__file__).parent / 'manifest.json'
        with open(manifest_path, 'r', encoding='utf-8') as f:
            manifest = json.load(f)

        required = ['tabs', 'scripting', 'clipboardWrite', 'activeTab']
        perms = manifest.get('permissions', [])
        missing = [p for p in required if p not in perms]

        if missing:
            TEST_RESULTS['warnings'].append("Missing permissions: " + ', '.join(missing))
            log_warn("Missing permissions: " + ', '.join(missing))

        if 'host_permissions' not in manifest:
            raise ValueError('Missing host_permissions')

        TEST_RESULTS['passed'].append('Permissions structure valid')
        log_success('Permissions validated')
    except Exception as e:
        TEST_RESULTS['failed'].append("Permission validation: " + str(e))
        log_error("Permission validation failed: " + str(e))

# ============================================
# Test 6: Icon validation
# ============================================
def test_icons():
    log_info('Test 6: Icon file validation')
    try:
        manifest_path = Path(__file__).parent / 'manifest.json'
        with open(manifest_path, 'r', encoding='utf-8') as f:
            manifest = json.load(f)

        icons = manifest.get('icons', {})
        if not icons or '128' not in icons:
            raise ValueError('Missing icon definition')

        icon_path = Path(__file__).parent / icons['128']
        if not icon_path.exists():
            raise ValueError("Icon file not found: " + icons['128'])

        TEST_RESULTS['passed'].append('Icon file present')
        log_success('Icon validated')
    except Exception as e:
        TEST_RESULTS['warnings'].append("Icon validation: " + str(e))
        log_warn("Icon validation: " + str(e))

# ============================================
# Test 7: JavaScript syntax check
# ============================================
def test_js_syntax():
    log_info('Test 7: JavaScript syntax validation')
    try:
        js_path = Path(__file__).parent / 'popup.js'
        with open(js_path, 'r', encoding='utf-8') as f:
            js = f.read()

        if js.count('{') != js.count('}'):
            raise ValueError("Mismatched braces")
        if js.count('(') != js.count(')'):
            raise ValueError("Mismatched parentheses")

        if 'chrome.tabs.create' not in js:
            raise ValueError("Missing chrome.tabs.create call")
        if 'chrome.scripting.executeScript' not in js:
            raise ValueError("Missing chrome.scripting.executeScript call")
        if 'navigator.clipboard.writeText' not in js:
            raise ValueError("Missing clipboard write functionality")

        TEST_RESULTS['passed'].append('JavaScript syntax valid')
        log_success('JavaScript syntax valid')
    except Exception as e:
        TEST_RESULTS['failed'].append("JS syntax: " + str(e))
        log_error("JS syntax check failed: " + str(e))

# ============================================
# Test 8: File integrity
# ============================================
def test_file_integrity():
    log_info('Test 8: File integrity validation')
    try:
        base_path = Path(__file__).parent
        required_files = [
            'manifest.json',
            'popup.html',
            'popup.js',
            'unlocked.png',
        ]

        missing = [f for f in required_files if not (base_path / f).exists()]
        if missing:
            raise ValueError("Missing files: " + ', '.join(missing))

        TEST_RESULTS['passed'].append('All required files present')
        log_success('File integrity valid')
    except Exception as e:
        TEST_RESULTS['failed'].append("File integrity: " + str(e))
        log_error("File integrity check failed: " + str(e))

# ============================================
# Main execution
# ============================================
def main():
    print('\n' + '='*60)
    print('Test: Discord Token Extractor Extension Validation')
    print('='*60 + '\n')

    # Run all tests
    test_manifest()
    test_popup_html()
    test_popup_js()
    test_token_extraction()
    test_permissions()
    test_icons()
    test_js_syntax()
    test_file_integrity()

    # Summary
    print('\n' + '='*60)
    print('Test Results Summary\n')

    if TEST_RESULTS['passed']:
        print('PASS (%d):' % len(TEST_RESULTS['passed']))
        for msg in TEST_RESULTS['passed']:
            print('   - ' + msg)

    if TEST_RESULTS['warnings']:
        print('\nWARNING (%d):' % len(TEST_RESULTS['warnings']))
        for msg in TEST_RESULTS['warnings']:
            print('   - ' + msg)

    if TEST_RESULTS['failed']:
        print('\nFAIL (%d):' % len(TEST_RESULTS['failed']))
        for msg in TEST_RESULTS['failed']:
            print('   - ' + msg)

    print('\n' + '='*60)
    total_tests = len(TEST_RESULTS['passed']) + len(TEST_RESULTS['failed'])
    if total_tests > 0:
        pass_rate = (len(TEST_RESULTS['passed']) / total_tests) * 100
        print('\nScore: %.1f%% (%d/%d)\n' % (pass_rate, len(TEST_RESULTS['passed']), total_tests))
    else:
        print('\nNo tests executed\n')

    return 0 if len(TEST_RESULTS['failed']) == 0 else 1

if __name__ == '__main__':
    sys.exit(main())

File: test-extension.js
/**
 * CTF: Discord Token Extractor 自動検証スクリプト
 * 拡張機能の各トークン抽出方法が正常に動作するかテスト
 */

const puppeteer = require('puppeteer');
const path = require('path');
const fs = require('fs');

const EXTENSION_PATH = path.resolve(__dirname);
const TEST_RESULTS = {
  passed: [],
  failed: [],
  warnings: [],
};

// ログ出力ユーティリティ
const log = {
  success: (msg) => console.log(`✅ ${msg}`),
  error: (msg) => console.error(`❌ ${msg}`),
  info: (msg) => console.log(`ℹ️  ${msg}`),
  warn: (msg) => console.warn(`⚠️  ${msg}`),
};

/**
 * Test 1: manifest.json の検証
 */
async function testManifest() {
  log.info('Test 1: manifest.json 構造検証');
  try {
    const manifest = JSON.parse(
      fs.readFileSync(path.join(EXTENSION_PATH, 'manifest.json'), 'utf8')
    );

    const required = ['manifest_version', 'name', 'version', 'action', 'permissions'];
    for (const field of required) {
      if (!manifest[field]) throw new Error(`Missing field: ${field}`);
    }

    if (manifest.manifest_version !== 3) {
      throw new Error('manifest_version must be 3');
    }

    if (!manifest.permissions.includes('scripting')) {
      throw new Error('Missing "scripting" permission');
    }

    if (!manifest.permissions.includes('clipboardWrite')) {
      throw new Error('Missing "clipboardWrite" permission');
    }

    if (!manifest.host_permissions?.some(p => p.includes('discord.com'))) {
      throw new Error('Missing Discord host permission');
    }

    TEST_RESULTS.passed.push('manifest.json structure valid');
    log.success('manifest.json valid');
  } catch (err) {
    TEST_RESULTS.failed.push(`Manifest validation: ${err.message}`);
    log.error(`manifest.json validation failed: ${err.message}`);
  }
}

/**
 * Test 2: popup.html の検証
 */
async function testPopupHTML() {
  log.info('Test 2: popup.html 構造検証');
  try {
    const html = fs.readFileSync(path.join(EXTENSION_PATH, 'popup.html'), 'utf8');

    if (!html.includes('extractButton')) throw new Error('Missing extractButton element');
    if (!html.includes('popup.js')) throw new Error('Missing popup.js script');
    if (!html.includes('Discord')) throw new Error('Missing Discord label text');

    TEST_RESULTS.passed.push('popup.html structure valid');
    log.success('popup.html valid');
  } catch (err) {
    TEST_RESULTS.failed.push(`HTML validation: ${err.message}`);
    log.error(`popup.html validation failed: ${err.message}`);
  }
}

/**
 * Test 3: popup.js の関数シグネチャ検証
 */
async function testPopupJS() {
  log.info('Test 3: popup.js 関数検証');
  try {
    const js = fs.readFileSync(path.join(EXTENSION_PATH, 'popup.js'), 'utf8');

    const required = ['extractTokenFromTab', 'extractToken'];
    for (const fn of required) {
      if (!js.includes(`function ${fn}`)) {
        throw new Error(`Missing function: ${fn}`);
      }
    }

    // 抽出方法の確認
    const methods = [
      { name: 'webpackChunk', pattern: /window\.webpackChunkdiscord_app/ },
      { name: 'localStorage', pattern: /localStorage\.getItem|localStorage\.token/ },
      { name: 'sessionStorage', pattern: /sessionStorage/ },
      { name: 'ReactFiber', pattern: /__react/ },
    ];

    const implemented = methods.filter(m => m.pattern.test(js));
    log.info(`Extraction methods implemented: ${implemented.map(m => m.name).join(', ')}`);

    if (implemented.length < 2) {
      TEST_RESULTS.warnings.push(`Only ${implemented.length} extraction methods found, expected 3+`);
      log.warn(`Only ${implemented.length} extraction methods found`);
    }

    TEST_RESULTS.passed.push('popup.js functions present and valid');
    log.success('popup.js valid');
  } catch (err) {
    TEST_RESULTS.failed.push(`JavaScript validation: ${err.message}`);
    log.error(`popup.js validation failed: ${err.message}`);
  }
}

/**
 * Test 4: ユニットテスト - トークン抽出関数の模擬実行
 */
async function testTokenExtraction() {
  log.info('Test 4: トークン抽出ロジック（模擬テスト）');
  try {
    // extractToken 関数を分離して実行可能に
    const mockWindow = {
      webpackChunkdiscord_app: [
        [Symbol()],
        {},
        (e) => {
          e.c = {
            test_module: {
              exports: {
                default: {
                  getToken: () => 'mock_token_from_webpack_12345',
                },
              },
            },
          };
        },
      ],
      localStorage: {
        getItem: () => null,
        token: undefined,
      },
      sessionStorage: {
        getItem: () => null,
        token: undefined,
      },
      document: {
        querySelectorAll: () => [],
      },
    };

    // 簡易的な抽出テスト
    let token = null;

    // webpack チャンク模擬
    for (const entry of mockWindow.webpackChunkdiscord_app) {
      if (typeof entry[2] === 'function') {
        entry[2]({
          c: {
            test: {
              exports: {
                default: { getToken: () => 'test_token_123' },
              },
            },
          },
        });
      }
    }

    if (!token && mockWindow.localStorage) {
      token = mockWindow.localStorage.getItem?.('token') || mockWindow.localStorage.token;
    }

    TEST_RESULTS.passed.push('Token extraction logic structure valid');
    log.success('Token extraction logic validated');
  } catch (err) {
    TEST_RESULTS.failed.push(`Token extraction test: ${err.message}`);
    log.error(`Token extraction test failed: ${err.message}`);
  }
}

/**
 * Test 5: パーミッション検証
 */
async function testPermissions() {
  log.info('Test 5: Chrome パーミッション検証');
  try {
    const manifest = JSON.parse(
      fs.readFileSync(path.join(EXTENSION_PATH, 'manifest.json'), 'utf8')
    );

    const required = ['tabs', 'scripting', 'clipboardWrite', 'activeTab'];
    const missing = required.filter(p => !manifest.permissions.includes(p));

    if (missing.length > 0) {
      TEST_RESULTS.warnings.push(`Missing permissions: ${missing.join(', ')}`);
      log.warn(`Missing permissions: ${missing.join(', ')}`);
    }

    if (!manifest.host_permissions) {
      throw new Error('Missing host_permissions');
    }

    TEST_RESULTS.passed.push('Permissions structure valid');
    log.success('Permissions validated');
  } catch (err) {
    TEST_RESULTS.failed.push(`Permission validation: ${err.message}`);
    log.error(`Permission validation failed: ${err.message}`);
  }
}

/**
 * Test 6: アイコン検証
 */
async function testIcons() {
  log.info('Test 6: アイコンファイル検証');
  try {
    const manifest = JSON.parse(
      fs.readFileSync(path.join(EXTENSION_PATH, 'manifest.json'), 'utf8')
    );

    const iconPath = manifest.icons?.[128];
    if (!iconPath) {
      throw new Error('Missing icon definition');
    }

    if (!fs.existsSync(path.join(EXTENSION_PATH, iconPath))) {
      throw new Error(`Icon file not found: ${iconPath}`);
    }

    TEST_RESULTS.passed.push('Icon file present');
    log.success('Icon validated');
  } catch (err) {
    TEST_RESULTS.warnings.push(`Icon validation: ${err.message}`);
    log.warn(`Icon validation: ${err.message}`);
  }
}

/**
 * メイン実行
 */
async function main() {
  console.log('\n🧪 Discord Token Extractor - 拡張機能検証テスト開始\n');
  console.log(`📁 拡張パス: ${EXTENSION_PATH}\n`);

  await testManifest();
  await testPopupHTML();
  await testPopupJS();
  await testTokenExtraction();
  await testPermissions();
  await testIcons();

  // 結果サマリー
  console.log('\n' + '='.repeat(50));
  console.log('📊 テスト結果サマリー\n');

  if (TEST_RESULTS.passed.length > 0) {
    console.log(`✅ 成功 (${TEST_RESULTS.passed.length}):`);
    TEST_RESULTS.passed.forEach(msg => console.log(`   • ${msg}`));
  }

  if (TEST_RESULTS.warnings.length > 0) {
    console.log(`\n⚠️  警告 (${TEST_RESULTS.warnings.length}):`);
    TEST_RESULTS.warnings.forEach(msg => console.log(`   • ${msg}`));
  }

  if (TEST_RESULTS.failed.length > 0) {
    console.log(`\n❌ 失敗 (${TEST_RESULTS.failed.length}):`);
    TEST_RESULTS.failed.forEach(msg => console.log(`   • ${msg}`));
  }

  console.log('\n' + '='.repeat(50));
  const totalTests = TEST_RESULTS.passed.length + TEST_RESULTS.failed.length;
  const passRate = ((TEST_RESULTS.passed.length / totalTests) * 100).toFixed(1);
  console.log(`\n📈 総合スコア: ${passRate}% (${TEST_RESULTS.passed.length}/${totalTests})\n`);

  const exitCode = TEST_RESULTS.failed.length === 0 ? 0 : 1;
  process.exit(exitCode);
}

main().catch(err => {
  console.error('Fatal error:', err);
  process.exit(1);
});
